### ISSUE TITLE
Missing ELF AArch32 relocation support in LLDB

### ISSUE THREAD

**Issue by weliveindetail** [llvm/llvm-project#61948]
LLVM JITLink gained [initial AArch32 support recently](5c1d160cd979dd6099f4266db600020a0cb4867a) and I started looking into the debugging story. I got JIT breakpoints on function names to work, but LLDB seems unable to show source information.

## Too long; didn't reproduce

LLDB's ELF object file plugin lacks relocation support for AArch32. In particular, it seems to consider all 32-bit targets to be i386:
https://github.com/llvm/llvm-project/blob/release/16.x/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp#L2697

## Step-by-step repro

I found the following results with release 16.0.0 (customized version from the [CodeLLDB plugin](https://github.com/vadimcn/codelldb)). Note that both, llvm-jitlink and llvm-jitlink-executor must be built from mainline.

(1) On the remote AArch32 device, run both lldb-server and llvm-jitlink-executor:
```console
> lldb-server version
lldb version 16.0.0-custom
  rust-enabled
> lldb-server platform --server --listen 0.0.0.0:9001 --gdbserver-port 9002 &
[1] 2710
> ./remote_build_main/bin/llvm-jitlink-executor listen=0.0.0.0:20000
```

(2) Connect LLDB to the remote lldb-server and let it attach to the llvm-jitlink-executor. Then set a breakpoint just before entering JITed code and resume execution:
```console
> lldb --version
lldb version 16.0.0-custom
  rust-enabled
> lldb
(lldb) platform select remote-linux
  Platform: remote-linux
 Connected: no
(lldb) platform connect connect://192.168.1.100:9001
  Platform: remote-linux
    Triple: arm-unknown-linux-gnueabihf
OS Version: 5.15.76 (5.15.76-v7+)
  Hostname: raspi32-3b
 Connected: yes
WorkingDir: /home/ez/.vscode-server/extensions/vadimcn.vscode-lldb-1.9.0/lldb
    Kernel: #1597 SMP Fri Nov 4 12:13:17 GMT 2022
(lldb) process attach --name llvm-jitlink-executor
Process 2835 stopped
* thread #1, name = 'llvm-jitlink-ex', stop reason = signal SIGSTOP
    frame #0: 0x76efc09c libpthread.so.0`accept + 36
libpthread.so.0`accept:
->  0x76efc09c <+36>: svc    #0x0
    0x76efc0a0 <+40>: cmn    r0, #4096
    0x76efc0a4 <+44>: mov    r4, r0
    0x76efc0a8 <+48>: bhi    0x76efc104                ; <+140>
Executable module set to "/home/ez/.lldb/module_cache/remote-linux/.cache/9B8F5E02-B8CD-9D55/llvm-jitlink-executor".
Architecture set to: arm-unknown-linux-gnueabihf.
(lldb) b TargetExecutionUtils.cpp:43
Breakpoint 1: where = llvm-jitlink-executor`llvm::orc::runAsMain(int (*)(int, char**), llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>, std::optional<llvm::StringRef>) + 708 at TargetExecutionUtils.cpp:43:3, address = 0x0012930c
(lldb) c
Process 2835 resuming
```

(3) JIT-link the example object to the remote executor:
```console
> host_build_main/bin/llvm-jitlink --oop-executor-connect=192.168.1.100:20000 base64char2val-minimal-debug.o
```

The debug session in (2) reports a number of errors before stopping at the breakpoint:
```console
error: JIT(0x76f67000)(0x0000000076f67000) unsupported 32-bit relocation: .rel.debug_info[0], type 2
error: JIT(0x76f67000)(0x0000000076f67000) unsupported 32-bit relocation: .rel.debug_info[1], type 2
...
error: JIT(0x76f67000)(0x0000000076f67000) unsupported 32-bit relocation: .rel.debug_info[30], type 2
error: JIT(0x76f67000)(0x0000000076f67000) unsupported 32-bit relocation: .rel.debug_line[0], type 2
Process 2835 stopped
* thread #4, name = 'llvm-jitlink-ex', stop reason = breakpoint 1.1
    frame #0: 0x0012930c llvm-jitlink-executor`llvm::orc::runAsMain(Main=(JIT(0x76f67000)`main + 1), Args=(Data = 0x00000000, Length = 0), ProgramName= Has Value=false ) at TargetExecutionUtils.cpp:43:3
   40  int RawArgc = Args.size() + !!ProgramName;
   41  char **RawArgv = ArgV.data();
   42
-> 43  printf("Calling int main(%d, 0x%08" PRIx64 ") @0x%08" PRIx64 "\n",
   44         RawArgc, pointerToJITTargetAddress(RawArgv),
   45         pointerToJITTargetAddress(Main));
   46  int Result = Main(RawArgc, RawArgv);
(lldb) b main
Breakpoint 2: 2 locations.
(lldb) br list
Current breakpoints:
1: file = 'TargetExecutionUtils.cpp', line = 43, exact_match = 0, locations = 1, resolved = 1, hit count = 1
  1.1: where = llvm-jitlink-executor`llvm::orc::runAsMain(int (*)(int, char**), llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>, std::optional<llvm::StringRef>) + 708 at TargetExecutionUtils.cpp:43:3, address = 0x0012930c, resolved, hit count = 1 

2: name = 'main', locations = 2, resolved = 2, hit count = 0
  2.1: where = llvm-jitlink-executor`main + 28 at llvm-jitlink-executor.cpp:120:35, address = 0x000dade4, resolved, hit count = 0 
  2.2: where = JIT(0x76f67000)`main, address = 0x76f6906a, resolved, hit count = 0 
```

Here we see that the JIT breakpoint 2.2 doesn't have source locations, while the breakpoint 2.1 in static code has them. So far I simply used GDB on AARch32, but since we have upstream support for AArch32 in JITLink now it's time to fix this.

**Comment by weliveindetail** [comment]
We could follow the example here: https://reviews.llvm.org/D145462

**Comment by weliveindetail** [comment]
Attaching the object file for step (3) [base64char2val-minimal-debug.o.zip](https://github.com/llvm/llvm-project/files/11159056/base64char2val-minimal-debug.o.zip)

**Comment by weliveindetail** [comment]
Confirming that this is fixed with D147642 on mainline:
```
(lldb) b main
Breakpoint 2: 2 locations.
(lldb) br list
Current breakpoints:
1: file = 'TargetExecutionUtils.cpp', line = 43, exact_match = 0, locations = 1, resolved = 1, hit count = 1
  1.1: where = llvm-jitlink-executor`llvm::orc::runAsMain(int (*)(int, char**), llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>, std::optional<llvm::StringRef>) + 708 at TargetExecutionUtils.cpp:43:3, address = 0x0012930c, resolved, hit count = 1 

2: name = 'main', locations = 2, resolved = 2, hit count = 0
  2.1: where = llvm-jitlink-executor`main + 28 at llvm-jitlink-executor.cpp:120:35, address = 0x000dade4, resolved, hit count = 0 
  2.2: where = JIT(0x76fb3000)`main + 14 at base64char2val.c:40:7, address = 0x76fb5078, resolved, hit count = 0 
```
And with appropriate source mapping:
```
...
(lldb) settings set target.source-map /remote/path /local/path
(lldb) s
(lldb) Process 23209 stopped
* thread #4, name = 'llvm-jitlink-ex', stop reason = step in
    frame #0: 0x76fb5098 JIT(0x76fb3000)`main(argc=0, argv=0x75800c68) at base64char2val.c:45:12
   42       return 1;
   43     }
   44  
-> 45     for (int i = 1; i < argc; i += 1) {
   46       printf("%s = ", argv[i]);
   47       for (int c = 0; argv[c]; c += 1)
   48         putchar(base64_Char2Val(argv[i][c]));
```
